cluster/engine.go

```diff
@@ -310,9 +310,11 @@ func (e *Engine) HealthIndicator() int64 {
 }
 
 // setState sets engine state
-func (e *Engine) setState(state engineState) {
-	e.Lock()
-	defer e.Unlock()
+func (e *Engine) setState(state engineState, lock bool) {
+	if lock {
```

Comment:
Like function `ValidationComplete`, it has already lock engine before `set state`, so we couldn't lock engine twice, it will case deadlock. In other case like function `CheckConnectionErr`, it doesn't lock engine before set state, so we should lock it first then set state. That's why we create the lock parameter.
@allencloud